### DISCUSSION TITLE
How to build gzrtp module with libzrtpcpp from Linux distribution?

### DISCUSSION THREAD

**Discussion by robert-scheck** [baresip/baresip#2234] [General]
[cmake/FindGZRTP.cmake](https://github.com/baresip/baresip/blob/82490ade7734a1ffd6cd10e85f12078e743080b7/cmake/FindGZRTP.cmake) tries to use `-lzrtpcppcore`, which isn't shipped by e.g. [Fedora 37](https://koji.fedoraproject.org/koji/rpminfo?rpmID=31276399).

**Comment by juha-h** [comment]
See https://github.com/baresip/baresip/blob/main/modules/gzrtp/module.mk.  Alternatively, you can clone from [https://github.com/juha-h/ZRTPCPP](url).

**Comment by robert-scheck** [comment]
[modules/gzrtp/module.mk](https://github.com/baresip/baresip/blob/main/modules/gzrtp/module.mk) also links against `-lzrtpcppcore` rather than `-lzrtpcpp` and requires a private copy of the library, why is that? From the packaging perspective of a Linux distribution, bundling 3rd-party libraries is usually somewhere between disliked and not permitted.

**Comment by juha-h** [comment]
> [modules/gzrtp/module.mk](https://github.com/baresip/baresip/blob/main/modules/gzrtp/module.mk) also links against `-lzrtpcppcore` rather than `-lzrtpcpp` and requires a private copy of the library, why is that? 

I think it is cause baresip does not need the full lib, i.e., client support.

> From the packaging perspective of a Linux distribution, bundling 3rd-party libraries is usually somewhere between disliked and not permitted.

I have not found ZRTPCPP lib in Debian, perhaps it is in some other distributions.

**Comment by robert-scheck** [comment]
According to [Repology](https://repology.org/project/libzrtpcpp/badges), ZRTPCPP is in various Linux distributions, however just as `libzrtpcpp.so(.X)` files not as `libzrtpcppcore.so(.X)` files. Do I get you right that you assume that linking against `-lzrtpcpp` from the Linux distribution rather than `-lzrtpcppcore` should work, too?

**Comment by juha-h** [comment]
Yes, full lib should be OK too.

**Comment by juha-h** [comment]
... but check that SDES in ON.  By default it is OFF: https://github.com/wernerd/ZRTPCPP/blob/6b3cd8e6783642292bad0c21e3e5e5ce45ff3e03/CMakeLists.txt#L39

**Comment by robert-scheck** [comment]
Why is SDES needed? And where is it actually used in the code? I didn't see any occurence of "sdes" in the code.

**Comment by juha-h** [comment]
Robert Scheck writes:

> Why is SDES needed? And where is it actually used in the code? I
> didn't see any occurence of "sdes" in the code.

In CMakeLists.txt there is this:

if (SDES AND NOT CCRTP)
    set (sdes_src ${CMAKE_SOURCE_DIR}/zrtp/ZrtpSdesStream.cpp)
endif()

So if SDES is OFF, this source file is not included.

I didn't write the module and don't know about the details.

**Comment by robert-scheck** [comment]
Wait, I meant where "sdes" would be used in "gzrtp" module in baresip…

**Comment by robert-scheck** [comment]
Why does the "gzrtp" module use internal zrtpcpp headers like `libzrtpcpp/ZRtp.h`?

**Comment by juha-h** [comment]
As I wrote in above, I didn't write gzrtp module.  I'm just using it and works fine with libzrtpcppcore.

The generated lib is called libzrtpcpp if CCRTP is ON:
```
if (CCRTP)
    set (PACKAGE libzrtpcpp)
    set(zrtplibName zrtpcpp)
```
By default CCRTP is OFF:
```
option(CCRTP "Build library to use with GNU ccRTP." OFF)
```
Try to find answers to your questions from ZRTPCPP  and gzrtp module sources.

**Comment by robert-scheck** [comment]
> As I wrote in above, I didn't write gzrtp module. I'm just using it and works fine with libzrtpcppcore.

I get this, but it's still strange that the gzrtp module in baresip uses non-public interfaces from libzrtpcpp, like `libzrtpcpp/ZRtp.h`. And I would like to understand why. Because even when building libzrtpcpp with `-DSDES=true -DCORE_LIB=true`, the result is:

  * `/usr/include/libzrtpcpp/ZrtpCWrapper.h`
  * `/usr/include/libzrtpcpp/ZrtpCallback.h`
  * `/usr/include/libzrtpcpp/ZrtpCodes.h`
  * `/usr/include/libzrtpcpp/ZrtpConfigure.h`
  * `/usr/include/libzrtpcpp/ZrtpUserCallback.h`
  * `/usr/include/libzrtpcpp/common/osSpecifics.h`
  * `/usr/lib64/libzrtpcppcore.so`
  * `/usr/lib64/libzrtpcppcore.so.4`
  * `/usr/lib64/libzrtpcppcore.so.4.6.6`
  * `/usr/lib64/pkgconfig/libzrtpcppcore.pc`
 
Maybe the author of the gzrtp module in baresip could comment here and explain this (odd) situation? I would expect that the gzrtp module in baresip uses only publicly available (as in via API and ABI exposed) libzrtpcpp interfaces rather than depending on libzrtpcpp internals.
